binderhub/utils.py

```diff
@@ -1,8 +1,37 @@
 """Miscellaneous utilities"""
 from collections import OrderedDict
+from hashlib import blake2b
+from math import log
+
 from traitlets import Integer, TraitError
 
 
+def blake2b_as_float(b):
+    """Compute Blake2 digest and convert it to a float.
+
+    Use this to create a float [0, 1) based on the Blake2 hash function.
+    """
+    return (
+        int.from_bytes(blake2b(b, digest_size=8).digest(), "big") / 0xFFFFFFFFFFFFFFFF
+    )
+
+
+def rendezvous_rank(buckets, key):
+    """Rank the buckets for a given key using Rendez-vous hashing
+
+    Each bucket is scored for the key and the best match is assigned the
+    highest score. The return value is a list of tuples of (score, bucket),
+    sorted in decreasing order.
+    """
+    ranking = []
+    for bucket in buckets:
+        h = blake2b_as_float(b"%s-%s" % (str(key).encode(), str(bucket).encode()))
```

Comment:
This is still missing here. I was hoping we could get this merged and tested in production before adding that.

The reason to have multiple "entries" for each bucket is to smooth out the clumpyness of random numbers as well as spread out assignments when a bucket is removed.

With (say) four buckets there is a high chance that one bucket receives a lot more than 25% of keys due to "clumpyness".

If we remove one of the four buckets then all of its keys will end up being assigned to one of the three remaining buckets, which might overload it.

The fix for both these problems is to represent each bucket with (say) 100 "pseudo buckets".

Code for the ranking would look something like:
```python
def rank(buckets, key, N=100):
    ranking = []
    for bucket in buckets:
        for i in range(N):
            score = blake2b_as_int(b"%s-%s-%s" % (str(key).encode(), str(bucket).encode(), str(i).encode()))
            ranking.append((score, bucket))

    return [b for (s, b) in sorted(ranking, reverse=True)]

Reply:
@betatim it is missing and pointless with this implementation, as we only generate bucket hashes with key as as a salt, as compared to generating hashes for the buckets in a standalone manner. We won't get the issues of clumping of many repos, because each bucket-repo combo will cause a new ranking where each bucket has an entirely different ranking order with entirely different scores.

Reply:
The issue where clumping arise due to the standalone hashes of individual bucket names, but won't arise if we always hash a "bucket name+repo name" string. Note that the "consistent hashing" algorithm was described like this and had that shortcoming, but the "rendezvous algorithm" in the linked slides you provided doesn't have this as it was described it was hashing "bucket name+repo name" as compared to consistent hashing that only hashed "bucket name".